src/com/sun/jna/Pointer.java

```diff
@@ -654,36 +656,110 @@ public ByteBuffer getByteBuffer(long offset, long length) {
+     * @param maxBytes
+     *            the maximum number of bytes to read
+     * @return the <code>String</code> value being pointed to, up to either a null
+     *         terminator or <code>maxBytes</code>
+     */
+    public String getWideString(long offset, int maxBytes) {
+        // Fetch the maxBytes
+        char[] data = this.getCharArray(offset, maxBytes / Native.WCHAR_SIZE);
+        // Convert to String using Wide String encoding
+        String ws = new String(data);
+        // Truncate at null if necessary
+        int indexOfNull = ws.indexOf('\0');
+        return indexOfNull < 0 ? ws : ws.substring(0, indexOfNull);
+    }
+
+    /**
+     * Copy native memory to a Java String. The encoding used is obtained form
+     * {@link Native#getDefaultStringEncoding()}.
+     *
+     * @param offset
+     *            byte offset from pointer to start reading bytes
      * @return the <code>String</code> value being pointed to
      */
     public String getString(long offset) {
         return getString(offset, Native.getDefaultStringEncoding());
     }
 
+    /**
+     * Copy native memory to a Java String. The encoding used is obtained form
+     * {@link Native#getDefaultStringEncoding()}.
+     *
+     * @param offset
+     *            byte offset from pointer to start reading bytes
+     * @param maxBytes
+     *            the maximum number of bytes to read
+     * @return the <code>String</code> value being pointed to, up to either a null
+     *         terminator or <code>maxBytes</code>
+     */
+    public String getString(long offset, int maxBytes) {
+        return getString(offset, maxBytes, Native.getDefaultStringEncoding());
+    }
+
     /**
      * Copy native memory to a Java String using the requested encoding.
      *
-     * @param offset byte offset from pointer to obtain the native string
-     * @param encoding the desired encoding
+     * @param offset
+     *            byte offset from pointer to obtain the native string
+     * @param encoding
+     *            the desired encoding
      * @return the <code>String</code> value being pointed to
      */
     public String getString(long offset, String encoding) {
         return Native.getString(this, offset, encoding);
     }
 
-    /** Read a native array of bytes of size <code>arraySize</code> from the
-        given <code>offset</code> from this {@link Pointer}.
-    */
+    /**
+     * Copy native memory to a Java String using the requested encoding.
+     *
+     * @param offset
+     *            byte offset from pointer to obtain the native string
+     * @param maxBytes
+     *            the maximum number of bytes to read
+     * @param encoding
+     *            the desired encoding
+     * @return the <code>String</code> value being pointed to, up to either a null
+     *         terminator or <code>maxBytes</code>
+     */
+    public String getString(long offset, int maxBytes, String encoding) {
```

Comment:
It should be pointed out, that this introduces a new potential issue for users, if `offset + maxBytes` is larger that the allocated memory, this will result in a segfault (if all goes well) and does a silent out-of-bounds read else.

Reply:
Hmm, good point.  I had actually thought at some point that I could pass `Integer.MAX_VALUE` here to make it behave "like before" without realizing that there would be an internal out-of-bounds read.  But when manipulating a `Pointer`, we don't know what the allocated memory is (and in fact the existing versions of all the `get*` and `read*` have the same flaw.)

A thought -- since I put these methods in primarily to enable checking bounds with the `Memory` class, what if these new methods were `protected`?

Reply:
Or as an alternative to `protected`, possibly just left the existing `Pointer` class alone and moved these "bounds checked" versions into the `Memory` class methods.

Reply:
Yes I think `Memory` would be a better place. 

However, one think was never addressed: @twall has a valid point, that an API that does not terminate a `char` array with a trailing `\0` is not a conventional C string. I'll get to the concrete sample `readlink`:

> readlink()  places  the  contents  of the symbolic link pathname in the buffer buf, which has size bufsiz.  readlink() does not append  a  null byte  to buf.  It will (silently) truncate the contents (to a length of bufsiz characters), in case the buffer is too small to hold all of  the contents.

So to use this correctly:
- the buffer that is used needs to be nulled before use
- you need to check the return value of the function if the returned number is equal to the buffer size, you need to do a reread with a larger buffer.

Just having a function that can read a C string without `\0` terminator is still broken.

So how often are these APIs encountered. The last time I saw something like that it was the windows registry and I think the solution as to allocate a buffer one char/byte larger than needed and get the null termination from there.

Reply:
I agree in general regarding "conventional C string".  Unfortunately I think the conversation so far did not address my specific use case as the authors of #106 and #1177 asked for a specified length for the String to handle embedded nulls.  I agree that when a fixed length is required, this need is better handled with a `ByteBuffer` or `byte[]` array as suggested.

As for how often this occurs, it may not be that common. Other than the `readlink()` example, I can't find any obvious ones on *nix platforms.  As far as Windows is concerned, I'm not completely sure, as it's not immediately clear if the "buffer too small" error messages result in the case where the result fits but the extra character does not, and I know of at least one case i've found where it doesn't, and the "allocate a buffer one char/byte larger than needed" may not protect against a race condition changing the string length by +1 between the length call and fetching the String.

One specific example is [CM_Get_Device_IDW](https://docs.microsoft.com/en-us/windows/win32/api/cfgmgr32/nf-cfgmgr32-cm_get_device_idw) has users call `CM_Get_Device_ID_Size` and add 1 before calling. See my contribution in #984 and our lengthy discussion about it.  The API notes "The function appends a NULL terminator to the supplied device instance ID string, unless the buffer is too small to hold the string. In this case, the function supplies as much of the identifier string as will fit into the buffer, and then returns CR_BUFFER_SMALL."   However, in the case where the string fits but the null doesn't, there is no error.  (And on re-reading I think my contribution is still broken for the exceedingly-rare-but-possible case of "string increasing by length of 1".)

There are a few other cases where similar "too small" errors are returned (and it's possible for the string to change in a race condition) and I haven't tested to see whether they would cause a problem. One example might be [https://docs.microsoft.com/en-us/windows/win32/api/processenv/nf-processenv-expandenvironmentstringsa](https://docs.microsoft.com/en-us/windows/win32/api/processenv/nf-processenv-expandenvironmentstringsa) although given the need for both a null terminator and an additional character I'm less concerned.

Bottom line, I agree that "Just having a function that can read a C string without \0 terminator is still broken."  It was less my intent to try to protect against broken APIs as to have all the Memory class's `get*()` methods "bounds check" safe.    

Since #1077 was still open as a "feature request" and #106 included a request to "Feel free to propose a full scenario, and explicitly indicate what parts you think should be handled by JNA (at either library or native level)." I decided to propose exactly the scenario I thought should be handled.  I'm totally fine if we close #1077 as. "won't fix" but include a bigger warning in the Memory class's `getString()` and `getWideString()` methods that it does not bounds check the end of the string (a warning hidden in a comment in the code).